Sources/FoundationNetworking/URLProtectionSpace.swift

```diff
@@ -332,19 +332,93 @@ open class URLProtectionSpace : NSObject, NSCopying {
 
 extension URLProtectionSpace {
     //an internal helper to create a URLProtectionSpace from a HTTPURLResponse 
-    static func create(using response: HTTPURLResponse) -> URLProtectionSpace {
+    static func create(with response: HTTPURLResponse) -> URLProtectionSpace? {
         let host = response.url?.host ?? ""
-        let port = response.url?.port ?? 80        //HTTP
-        let _protocol = response.url?.scheme
-        guard let wwwAuthHeader = response.allHeaderFields["Www-Authenticate"] as? String else {
-            fatalError("Authentication failed but no Www-Authenticate header in response")
+
+        let scheme = response.url?.scheme
+
+        let port: Int = {
+            if let port = response.url?.port {
+                return port
+            }
+
+            switch scheme {
+            case "http":
+                return 80
+            case "https":
+                return 443
+            default:
+                return 80
+            }
+        }()
+
+        // Assuming HTTPURLResponse does capitalization of standard headers
+        guard let authenticateValue = response.allHeaderFields["Www-Authenticate"] as? String else {
+            return nil
         }
 
-        //The format of the authentication header is `<auth-scheme> realm="<realm value>"`
-        //Example: `Basic realm="Fake Realm"`
-        let authMethod = wwwAuthHeader.components(separatedBy: " ")[0]
-        let realm = String(String(wwwAuthHeader.components(separatedBy: "realm=")[1].dropFirst()).dropLast())
-        return URLProtectionSpace(host: host, port: port, protocol: _protocol, realm: realm, authenticationMethod: authMethod)
+        // Typical Www-Authenticate header is something like
+        //   Www-Authenticate: Digest realm="test", domain="/HTTP/Digest", nonce="e3d002b9b2080453fdacea2d89f2d102"
+
+        let components = authenticateValue.components(separatedBy: " ")
+        let authMethod: String? = {
+            guard let authMethod = components.first else {
+                return nil
+            }
+
+            switch authMethod.lowercased() {
+            case "basic":
+                return NSURLAuthenticationMethodHTTPBasic
+            case "digest":
+                return NSURLAuthenticationMethodHTTPDigest
+            case "ntlm":
+                return NSURLAuthenticationMethodNTLM
+            case "negotiate":
+                return NSURLAuthenticationMethodNegotiate
+            default:
```

Comment:
Will it cause problems to create a `URLProtectionSpace` for methods that are not currently implemented (ie everything except Basic)? Could always return `nil` for these other ones for now

Reply:
I actually thought about this helper point as parsing as much info as possible (and reasonable). Hmm, but on the other side it could take responsibility to cut off unsupported things. Agree, this should return nil for everything other than Basic auth.